src/JsonSchema/Validator.php

```diff
@@ -43,13 +41,50 @@ public function check($value, $schema = null, JsonPointer $path = null, $i = nul
      * types defined in the schema whenever possible. Note that the first argument is passed by reference,
      * so you must pass in a variable.
      *
+     * This method is now a simple alias, and exists only so that the API is not fundamentally broken for
+     * people who use this. This method should probably be officially deprecated at some point.
+     *
      * {@inheritDoc}
      */
-    public function coerce(&$value, $schema = null, JsonPointer $path = null, $i = null)
+    public function coerce(&$value, $schema = null)
     {
-        $validator = $this->factory->createInstanceFor('schema');
-        $validator->coerce($value, $schema);
+        // Preserve the current checkMode, because users of this API expect to be
+        // able to set it via the factory before calling this method.
+        $checkMode = $this->factory->getCheckMode() | self::CHECK_MODE_COERCE;
 
-        $this->addErrors(array_unique($validator->getErrors(), SORT_REGULAR));
+        return $this->validate($value, $schema, $checkMode);
+    }
+
+    /**
+     * Validates the given data against the schema and returns an object containing the results
+     * Both the php object and the schema are supposed to be a result of a json_decode call.
+     * The validation works as defined by the schema proposal in http://json-schema.org.
+     *
+     * @param mixed $value Decoded JSON document to validate
+     * @param mixed $schema Decoded JSON schema to validate against
+     * @param int $checkMode Set various options to change the validation behavior
+     *
+     * {@inheritDoc}
+     */
+    public function validate(&$value, $schema = null, $checkMode = self::CHECK_MODE_NORMAL)
```

Comment:
Hmm.... maybe we see this a little differently? I don't feel like this is a corner.

We went from two public methods (`check` and `coerce`) to two public methods (`validate` and `validateImmutable`), and in doing so significantly improved the interface from an end-user perspective, plus allowing for easy addition of new functionality in the future without needing to change the public API again.

If you don't care about pass-by-value at all, then `validateImmutable` can be removed; that's the entire reason it's there.

If you don't care about breaking backwards-compatibility, then `check` and `coerce` can certainly be dropped... but IMO keeping them is safer - they're just aliases now anyway. Perhaps having them trigger a deprecation notice when called would be a sensible middle-ground, and then remove them in a future release?

`_check` is *not* a public method, whatever its visibility keyword says. The only reason it exists at all is because `ConstraintInterface` requires it. Would making `_check` be a noop be an improvement? Because the content of that method can easily be moved into `validate`.